### DISCUSSION TITLE
Frame rendering overwrites explicitly set cursor

### DISCUSSION THREAD

**Discussion by vvurzelsepp** [ratatui/ratatui#1327] [Q&A] [answered]
Hi, 
I might be doing something wrong here, but I can't figure it out. I wrote an `Input` component which implements the `Widget` trait. As the name might suggest, it's a text input and needs to set the current cursor position. Now the `render` function only receives a `Rect` and `Buffer`, but no `Frame` object, so I can't access `set_cursor` / `set_cursor_position` directly. 
I don't want to control the cursor from the outside, at least not directly, so I tried to use crossterms `cursor` module:

```rust
// inside
// impl Widget for Input
// fn render() { ... }
let _ = crossterm::execute!(
    std::io::stdout(),
    crossterm::cursor::Show,
    crossterm::cursor::MoveTo(column, row),
    crossterm::cursor::SetCursorStyle::BlinkingBlock,
);
```
This only works for the fraction of a second: If I type fast enough, I can see the cursor visually for a super short period of time. And after that, the cursor disappears from the screen. It seems like the rendering process is overwriting my crossterm cursor settings. How can I control this? 

I am using
ratatui.rs v0.28.0
crossterm.rs v0.28.1

Greetings

**Comment by joshka** [comment, marked as answer]
The terminal draw method moves the cursor around to draw the buffer. Any crossterm stuff you do outside of this is ignored. For the most part Ratatui assumes that it's in control of what's happening to the display and trying to do anything outside will lead to weird results.

From a app hierarchy perspective, there's only one cursor location, so allowing widgets to set the cursor doesn't make much sense. We've talked briefly about adding the cursor to the buffer, but I don't think that's a good idea in general either. It has to be something that is outside of the widget that handles setting the cursor position. So you'll need to communicate the cursor position back to the code that draws your widget somehow. It's likely the position is only known after you've drawn the widget (because the widget can be drawn anywhere on the screen, and the screen size can be any size).

There's a few different ways to do this:
1. Implement `StatefulWidget for Input` and have a cursor position in the state. After rendering the widget, call frame.cursor_position()
2. Implement `Widget for &mut Input` and have a cursor position in the Input. After rendering the widget, call frame.cursor_position()
3. Use channels or shared memory to communicate cursor changes with your app. After drawing, call terminal.cursor_position()

I'd choose 1 if you want a lot of docs and previous code to work with (downside is more code). Choose 2 if you want less examples, due to this being a bit of a newer approach, but something that works a nicely and is simple-ish. Choose 3 if you have many widgets which can control the cursor (e.g. if each instance tracks whether it's focused or not, then sends cursor change messages / changes the cursor in a shared mutable value as needed).

**Reply by vvurzelsepp** [reply]
Hi joshka,
thanks for the explanation. It makes sense to me now, why this happens. It's a little bit unfortunate that this behavior and the solutions to it are not documented. Especially for a newbie it's not exactly clear how to handle this problem.

If you need any help with documentation and examples I could try and prepare something. Obviously, it has to be checked due to my limited experience with ratatui and tui in general. 😅

For my project I have to go with option 3, since I have frames with multiple input components.

**Reply by orhun** [reply]
We always welcome new contributions to the [website](https://github.com/ratatui-org/ratatui-website) so please go ahead with it! Updating an example (or adding a new one e.g. `cursor.rs`) is also a very neat idea :)
